### ISSUE TITLE
formatter(Jupyter): ending semicolons valid in Jupyter

### ISSUE THREAD

**Issue by henryiii** [astral-sh/ruff#8254]
The Jupyter formatter has a couple of issues when running on notebooks.

One is it removes semicolons at the end of lines. `black[jupyter]` does not remove semicolons at the end of lines when running on a notebook, since it's used to mark output as hidden, for example when plotting. `plt.plot();` will show the plot but not print out the repr for the Artists that the plot method returns.

Edit: On the last line of a cell. Related: #7300 which is for the linter. But that's rule based, so it's easy to turn it off for `*.ipynb` files, while there isn't a workaround for the formatter, making it a blocker.

(Fixed in main) The other is that it can't read a notebook that black and ruff-lint are able to read. The error is:

```
error: Failed to format docs/examples/HistDemo.ipynb: source contains syntax errors: ParseError { error: UnrecognizedToken(Percent, None), offset: 0, source_path: "<filename>" }
```

I'm guessing it's choking on a line magic, perhaps? The file does have this line:

```python
%config InteractiveShell.ast_node_interactivity="last_expr_or_assign"
```

Which is the only `%` in the file.

**Comment by charliermarsh** [comment]
Thanks!

**Comment by dhruvmanila** [comment]
Re line magic, it's solved https://github.com/astral-sh/ruff/issues/8204, sorry for the churn!

**Comment by dhruvmanila** [comment]
Related: https://github.com/astral-sh/ruff/issues/7300

**Comment by MichaReiser** [comment]
Is my assumption correct that this only applies to expression statements or are there other statements for which semicolons need to be preserved?

**Comment by henryiii** [comment]
IPython has several modes. If you set:

```
%config InteractiveShell.ast_node_interactivity="last_expr_or_assign"
```

Then either `val = expr` or just `expr` will print if it's the final statement in the cell, unless there's a semicolon (see, for example, https://hist.readthedocs.io/en/latest/examples/HistDemo.html). By default, only `expr` will display, and `val = expr` won't. I'd say simply allowing a final semicolon in cells would be fine.

IIRC `a.b =` and `a[b] =` statements don't display even in the enhanced mode, but could be wrong.

**Comment by MichaReiser** [comment]
Note, there's now a helper function to extract the trailing semicolon. We can use it in the assignment (including augmented and annotated) and expression statement formatting

https://github.com/astral-sh/ruff/blob/2c84f911c4d57f3b436458f4d0e34fe10655cf76/crates/ruff_python_formatter/src/statement/mod.rs#L89-L107

**Comment by dhruvmanila** [comment]
Black's implementation for reference: https://github.com/psf/black/blob/c54c213d6a3132986feede0cf0525f5bae5b43d6/src/black/handle_ipynb_magics.py#L73-L102. It seems to be removing them for all cases.

**Comment by henryiii** [comment]
The function after that puts them back - not sure why you'd need to remove it and then put it back, since the normal formatter also removes them - just checking to see if it's there and then putting it back seems like it would also work. Anyway, what cases are not covered by assignments and expressions? Blocks can't end lines, for example. Simicolons aren't needed for import statements - I wonder if black keeps them, actually. :)

**Comment by MichaReiser** [comment]
> Black's implementation for reference: [psf/black@`c54c213`/src/black/handle_ipynb_magics.py#L73-L102](https://github.com/psf/black/blob/c54c213d6a3132986feede0cf0525f5bae5b43d6/src/black/handle_ipynb_magics.py#L73-L102). It seems to be removing them for all cases.

Nice find. Removing them and adding them back sounds complicated. We can just preserve them 😄
